design/crypto_primitives.rs

```diff
@@ -0,0 +1,113 @@
+//! `PublicKeys` and `SecretKeys` into corresponding types: `PublicEncryptKey, `SecretEncryptKey`,
+//! `PublicSignKey` and `SecretSignKey`.
+
+// no signing key
+pub struct PublicEncryptKey {
+    encrypt: box_::PublicKey,
+}
+
+impl PublicEncryptKey {
+    // No new() method, since public key should be constructed using `SecretEncryptKey`
+    // pub fn new() -> Self
+
+    /// Construct public key from bytes. Useful when it was serialized before.
+    pub fn from_bytes(public_key: &[u8]) -> Self
+
+    /// For anyone who wants to store public key.
+    pub fn as_bytes(&self) -> &[u8]
+
+    pub fn encrypt_anonymous<T: Serialize>(&self, plaintext: &T) -> Result<Vec<u8>, Error>
+
+    pub fn encrypt_anonymous_bytes(&self, plaintext: &[u8]) -> Vec<u8> {
+}
+
+/// Store secret key in `Arc` to discourage cloning it. Makes it more secure by not duplicating
+/// encryption keys in memory.
+pub struct SecretEncryptKey {
+    inner: Arc<SecretEncryptKeyInner>,
+    public: PublicKey,
+}
+
+struct SecretEncryptKeyInner {
+    encrypt: box_::SecretKey,
+}
+
+impl SecretEncryptKey {
+    /// Construct random public/private encryption key pair.
+    pub fn new() -> Self
+
+    /// Construct secret key from given bytes. Useful when secret key was serialized before.
+    // Note that we don't need to provide public key, seems like it's derived from secret key:
+    // https://docs.rs/rust_sodium/0.9.0/rust_sodium/crypto/box_/curve25519xsalsa20poly130/struct.SecretKey.html#method.public_key
+    pub fn from_bytes(secret_key: &[u8]) -> Self
+
+    /// Returns the public part of the secret key set.
+    pub fn pub_key(&self) -> &PublicEncryptKey
+
+    /// Derive shared key for symmetric data encryption
+    pub fn shared_secret(&self, their_pk: &PublicEncryptKey) -> SharedSecretKey
+
+    /// For anyone who wants to store secret key.
+    pub fn as_bytes(&self) -> &[u8]
+
+    pub fn decrypt_anonymous<T: Serialize + DeserializeOwned>(&self, ciphertext: &[u8]) -> Result<T, Error>
+
+    pub fn decrypt_anonymous_bytes(&self, ciphertext: &[u8]) -> Result<Vec<u8>, Error>
+}
+
+pub struct PublicSignKey {
+    sign: sign::PublicKey,
+}
+
+impl PublicSignKey {
+    pub fn verify_detached(&self, signature: &Signature, data: &[u8]) -> bool {
+
+    pub fn from_bytes(public_key: &[u8]) -> Self
+
+    pub fn as_bytes(&self) -> &[u8]
+}
+
+pub struct SecretSignKey {
+    inner: Arc<SecretSignKeyInner>,
+    public: PublicSignKey,
+}
+
+struct SecretSignKeyInner {
+    encrypt: sign::SecretKey,
+}
+
+impl SecretSignKey {
+    /// Construct random public/private signing key pair.
+    pub fn new() -> Self
+
+    pub fn from_bytes(secret_key: &[u8]) -> Self
+
+    pub fn as_bytes(&self) -> &[u8]
+
+    pub fn sign_detached(&self, data: &[u8]) -> Signature {
+}
+
+
+//! Types that remain intact as in https://github.com/maidsafe/safe_crypto/blob/f214410/src/lib.rs:
+//! SharedSecretKey
+//! Signature
+//! SymmetricKey
```

Comment:
just like above for shared secret, there's no facility to obtain raw() like other secret keys. Better to add that too.